### ISSUE TITLE
Raft catch-up logic is not efficient

### ISSUE THREAD

**Issue by jumaffre** [microsoft/CCF#1451]
When a new node joins a network, it currently has to replay all transactions since the beginning of time (for a way to do this more effectively, see https://github.com/microsoft/CCF/milestone/12). However, the current catch up logic seems inefficient in many ways (with primary's node log evidence for each claim):
- The first Append Entries message sent to the new joiner is sent from the primary's `last_idx`, which will automatically be rejected by the backup:
```
2020-07-28T13:55:25.602905Z -0.554 0   [info ] ../src/node/channels.h:344           | node channel with 9 is now established
...
2020-07-28T13:55:25.664574Z -0.436 0   [debug] ../src/consensus/raft/raft.h:499     | Send append entries from 8 to 9: 155 to 154 (152)
2020-07-28T13:55:25.664684Z        100 [debug] ../src/host/node_connections.h:278   | send AE to node 9 [96]: 154, 154
2020-07-28T13:55:25.664762Z -0.437 0   [debug] ../src/consensus/raft/raft.h:785     | Recv append entries response to 8 from 9: failed
```
- Further Append Entries are sent one by one:
```
2020-07-28T13:55:25.664843Z -0.437 0   [debug] ../src/consensus/raft/raft.h:499     | Send append entries from 8 to 9: 1 to 2 (152)
2020-07-28T13:55:25.686221Z        100 [debug] ../src/host/node_connections.h:278   | send AE to node 9 [17255]: 2, 0
2020-07-28T13:55:25.687620Z -0.459 0   [debug] ../src/consensus/raft/raft.h:499     | Send append entries from 8 to 9: 3 to 3 (152)
2020-07-28T13:55:25.688096Z        100 [debug] ../src/host/node_connections.h:278   | send AE to node 9 [1088]: 3, 2
2020-07-28T13:55:25.688246Z -0.460 0   [debug] ../src/consensus/raft/raft.h:499     | Send append entries from 8 to 9: 4 to 4 (152)
2020-07-28T13:55:25.688913Z        100 [debug] ../src/host/node_connections.h:278   | send AE to node 9 [1675]: 4, 3
2020-07-28T13:55:25.689093Z -0.461 0   [debug] ../src/consensus/raft/raft.h:499     | Send append entries from 8 to 9: 5 to 5 (152)
2020-07-28T13:55:25.689470Z        100 [debug] ../src/host/node_connections.h:278   | send AE to node 9 [132]: 5, 4
...
2020-07-28T13:55:27.362209Z        100 [debug] ../src/host/node_connections.h:278   | send AE to node 9 [11085]: 152, 151
2020-07-28T13:55:27.363345Z -2.135 0   [debug] ../src/consensus/raft/raft.h:499     | Send append entries from 8 to 9: 153 to 153 (152)
2020-07-28T13:55:27.364034Z        100 [debug] ../src/host/node_connections.h:278   | send AE to node 9 [1092]: 153, 152
2020-07-28T13:55:27.364178Z -2.136 0   [debug] ../src/consensus/raft/raft.h:499     | Send append entries from 8 to 9: 154 to 154 (152)
2020-07-28T13:55:27.369103Z        100 [debug] ../src/host/node_connections.h:278   | send AE to node 9 [11183]: 154, 153
```
- Because of in-flight AE, it is possible that the leader sends a batch of AEs twice, even the first batch was deserialised successfully by the backup:
```
2020-07-28T13:55:27.364178Z -2.136 0   [debug] ../src/consensus/raft/raft.h:499     | Send append entries from 8 to 9: 154 to 154 (152) <<< This entry and everything before was successfully deserialised by the backup
2020-07-28T13:55:27.369103Z        100 [debug] ../src/host/node_connections.h:278   | send AE to node 9 [11183]: 154, 153
2020-07-28T13:55:27.370009Z -2.142 0   [debug] ../src/consensus/raft/raft.h:785     | Recv append entries response to 8 from 9: failed <<< Some in-flight AE telling the primary that the backup didn't accept *some* AE
2020-07-28T13:55:27.370058Z -2.142 0   [debug] ../src/consensus/raft/raft.h:499     | Send append entries from 8 to 9: 1 to 2 (152)  <<< Sending everything *again*
2020-07-28T13:55:27.393098Z        100 [debug] ../src/host/node_connections.h:278   | send AE to node 9 [17255]: 2, 0
2020-07-28T13:55:27.394514Z -2.166 0   [debug] ../src/consensus/raft/raft.h:499     | Send append entries from 8 to 9: 3 to 3 (152)
2020-07-28T13:55:27.395192Z        100 [debug] ../src/host/node_connections.h:278   | send AE to node 9 [1088]: 3, 2
...
2020-07-28T13:55:29.051999Z -3.824 0   [debug] ../src/consensus/raft/raft.h:499     | Send append entries from 8 to 9: 154 to 154 (152)
2020-07-28T13:55:29.073168Z        100 [debug] ../src/host/node_connections.h:278   | send AE to node 9 [11183]: 154, 153
```

And again:
```
2020-07-28T13:55:29.097997Z -3.843 0   [debug] ../src/consensus/raft/raft.h:785     | Recv append entries response to 8 from 9: failed
2020-07-28T13:55:29.098059Z -3.843 0   [debug] ../src/consensus/raft/raft.h:499     | Send append entries from 8 to 9: 1 to 2 (152)
2020-07-28T13:55:29.106889Z        100 [debug] ../src/host/node_connections.h:278   | send AE to node 9 [17255]: 2, 0
...
2020-07-28T13:55:30.967672Z -5.713 0   [debug] ../src/consensus/raft/raft.h:499     | Send append entries from 8 to 9: 154 to 154 (152)
2020-07-28T13:55:30.988085Z        100 [debug] ../src/host/node_connections.h:278   | send AE to node 9 [11183]: 154, 153
```

Before we finally get some successfully responses from the backup:
```
2020-07-28T13:55:30.989239Z -5.734 0   [debug] ../src/consensus/raft/raft.h:794     | Recv append entries response to 8 from 9 for index 2: success
2020-07-28T13:55:30.989297Z -5.734 0   [debug] ../src/consensus/raft/raft.h:1113    | In update_commit, new_commit_idx: 2, last_idx: 154
```

Note that it took 5 seconds (!) for the primary to receive the first successful Append Entries message from the new joiner. I'm attaching the primary's log to this issue (taken from a verbose run of the `full_test_suite`). 

As discussed with @eddyashton, this also causes an unresponsive HTTPS server as no RPC connections can be created on the host while these Append Entries messages are being processed by the primary's host from the ring buffer. 

**Attachments:**
- [primary_log.txt](https://github.com/microsoft/CCF/files/4989283/primary_log.txt)

**Comment by achamayou** [comment]
@jumaffre is this still a serious concern with recent changes to fairness and the addition of snapshotting?

**Comment by achamayou** [comment]
This should be addressed by #3151, other efficiency improvements for Raft can be filed separately.
